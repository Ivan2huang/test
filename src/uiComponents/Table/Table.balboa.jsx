import React from 'react';
import { useTheme } from '@material-ui/core';
import Table from './Table';

const TableContainer = props => {
  const theme = useTheme();
  return (
    <Table
      customStyles={{
        borderTop: `1px solid ${theme.disabled}`,
        borderLeft: `1px solid ${theme.disabled}`,
        borderRight: `1px solid ${theme.disabled}`,
        borderBottom: `1px solid ${theme.disabled}`,
        headerBackground: theme.grey1,
        headerTextType: 'style-6',
      }}
      {...props}
    />
  );
};

export default TableContainer;
