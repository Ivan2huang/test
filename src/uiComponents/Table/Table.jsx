import React from 'react';
import PropTypes from 'prop-types';

import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
  makeStyles,
} from '@material-ui/core';

import Typography from '../Typography';
import { onEnter } from '../../helpers/helpers';
import { buildMap } from '../../modules/benefits/benefits/helper';

const useStyles = makeStyles(theme => ({
  tableRowHover: {
    cursor: 'pointer',
    '&:focus': {
      background: theme.grey1,
    },
  },
  tableCellRoot: ({ borderLeft, borderRight, borderBottom }) => ({
    borderLeft,
    borderRight,
    borderBottom,
    verticalAlign: 'top',
    paddingTop: theme.spacingX(3),
    paddingBottom: theme.spacingX(3),
  }),
  tableHeaderCell: ({ borderTop, headerBackground }) => ({
    borderTop,
    background: headerBackground,
  }),
}));

const StyledTable = ({
  columnDefs,
  data,
  hover,
  groupBy,
  onRowClick,
  customStyles,
  rowTabIndex,
}) => {
  const classes = useStyles(customStyles);
  const { headerTextType } = customStyles;
  const groupByKey = groupBy || columnDefs[0].name;
  const map = buildMap(data, groupByKey);
  let previousRenderedRow = '';

  const renderTableHeader = () =>
    columnDefs.map((column, index) => {
      const id = `table-header-${index}`;
      return (
        <TableCell
          key={id}
          align={column.alignContent}
          classes={{
            root: classes.tableCellRoot,
            head: classes.tableHeaderCell,
          }}
        >
          <Typography type={headerTextType}>{column.name}</Typography>
        </TableCell>
      );
    });

  const getTableCell = (key, column, row, rowSpan) => (
    <TableCell
      key={key}
      align={column.alignContent}
      rowSpan={rowSpan}
      classes={{ root: classes.tableCellRoot }}
    >
      <Typography type="style-6">{column.template(row)}</Typography>
    </TableCell>
  );

  const renderTableCells = (row, rowIndex) =>
    columnDefs.map((column, index) => {
      const key = `table-cell-${rowIndex}-${index}`;
      if (column.pivotColumn && row[groupByKey] !== previousRenderedRow) {
        previousRenderedRow = row[groupByKey];
        const withRowSpan = map[row[groupByKey]];
        return getTableCell(key, column, row, withRowSpan);
      }
      if (column.pivotColumn && row[groupByKey] === previousRenderedRow) {
        return undefined;
      }
      return getTableCell(key, column, row);
    });

  const renderTableRows = rows =>
    rows.map((row, rowIndex) => {
      const key = `table-row-${rowIndex}`;
      return (
        <TableRow
          classes={{ hover: classes.tableRowHover }}
          hover={hover}
          onClick={() => onRowClick(row)}
          key={key}
          tabIndex={rowTabIndex}
          onKeyPress={e => onEnter(e, () => onRowClick(row))}
        >
          {renderTableCells(row, rowIndex)}
        </TableRow>
      );
    });

  return (
    <Table>
      <TableHead>
        <TableRow>{renderTableHeader()}</TableRow>
      </TableHead>
      <TableBody>{renderTableRows(data)}</TableBody>
    </Table>
  );
};

StyledTable.defaultProps = {
  hover: false,
  groupBy: undefined,
  onRowClick: () => {},
  customStyles: {
    borderTop: 'none',
    borderLeft: 'none',
    borderRight: 'none',
    borderBottom: 'none',
    headerBackground: 'none',
    headerTextType: 'style-8',
  },
};

StyledTable.propTypes = {
  columnDefs: PropTypes.arrayOf(
    PropTypes.exact({
      name: PropTypes.string.isRequired,
      template: PropTypes.func.isRequired,
      alignContent: PropTypes.string,
      pivotColumn: PropTypes.bool,
    }),
  ).isRequired,
  data: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  hover: PropTypes.bool,
  groupBy: PropTypes.string,
  onRowClick: PropTypes.func,
  customStyles: PropTypes.shape({}),
  rowTabIndex: PropTypes.string,
};

StyledTable.defaultProps = {
  rowTabIndex: '0',
  onRowClick: () => {},
};

export default StyledTable;
