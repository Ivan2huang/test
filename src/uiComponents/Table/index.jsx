import CONFIG from '../../constants/config';
import defaultTable from './Table.default';
import BalboaTable from './Table.balboa';

const containers = {
  balboa: BalboaTable,
};

const container = containers[CONFIG.themeCode] || defaultTable;
export default container;
