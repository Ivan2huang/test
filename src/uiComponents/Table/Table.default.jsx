import React from 'react';
import { useTheme } from '@material-ui/core';
import Table from './Table';

const TableContainer = props => {
  const theme = useTheme();
  return (
    <Table
      customStyles={{
        borderTop: 'none',
        borderLeft: 'none',
        borderRight: 'none',
        borderBottom: `1px solid ${theme.grey1}`,
        headerBackground: 'none',
        headerTextType: 'style-8',
      }}
      {...props}
    />
  );
};

export default TableContainer;
