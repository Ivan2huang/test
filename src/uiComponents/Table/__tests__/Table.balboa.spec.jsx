/* eslint-disable react/prop-types */
import React from 'react';
import { render } from '@testing-library/react';

import Table from '../Table.balboa';
import withTheme from '../../../themes/withThemeProvider';

jest.mock(
  '../Table',
  () => ({
    customStyles: {
      borderTop,
      borderLeft,
      borderRight,
      borderBottom,
      headerTextType,
      headerBackground,
    },
  }) => (
    <table
      style={{
        borderTop,
        borderLeft,
        borderRight,
        borderBottom,
        headerBackground,
        headerTextType,
      }}
    />
  ),
);
jest.mock('../../../constants/config', () => jest.fn());

describe('Table component', () => {
  it('should match the snapshot', () => {
    const Component = withTheme(Table);
    const { container } = render(<Component />);

    expect(container).toMatchSnapshot();
  });
});
