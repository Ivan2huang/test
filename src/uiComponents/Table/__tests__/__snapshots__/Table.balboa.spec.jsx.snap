// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component should match the snapshot 1`] = `
<div>
  <table
    style="border-top: 1px solid #D7D8D6; border-left: 1px solid #D7D8D6; border-right: 1px solid #D7D8D6; border-bottom: 1px solid #D7D8D6;"
  />
</div>
`;
